File: 2021/src/day2.clj
(ns day2 (:require [clojure.string :refer [split-lines split]]))

(def input (-> "./input/day2input.txt"
               clojure.core/slurp
               split-lines
               (as-> t (map #(split % #" ") t))
               (as-> t (into [] t))))

(def part1
  (reduce
   (fn [position [movement value]]
     (let [iValue (Integer/parseInt value)]
       (condp = movement
         "forward" (update position :horizontal + iValue)
         "down" (update position :depth + iValue)
         "up" (update position :depth - iValue))))
   {:horizontal 0 :depth 0} input))
 
(def part2
  (reduce
   (fn [position [movement value]]
     (let [iValue (Integer/parseInt value)]
       (condp = movement
         "forward" (-> position
                       (update :horizontal + iValue)
                       (update :depth + (* iValue (:aim position))))
         "down" (update position :aim + iValue)
         "up" (update position :aim - iValue))))
   {:horizontal 0 :depth 0 :aim 0} input))

File: 2021/src/day2.fs
module day2
open FSharpx.Text

type foldState =
    { horizontal: int
      depth: int
      aim: int }

let input =
    "./input/day2input.txt"
    |> System.IO.File.ReadAllLines
    |> Array.map (Strings.split ' ')
    |> Array.map (fun x -> (x.[0], System.Int32.Parse(x.[1])))

let p1 =
    input
    |> Array.fold
        (fun acc value ->
            match fst value with
            | "forward" -> { acc with horizontal = acc.horizontal + snd value }
            | "down" -> { acc with depth = acc.depth + snd value }
            | "up" -> { acc with depth = acc.depth - snd value }
            | _ -> acc)

        { horizontal = 0; depth = 0; aim = 0 }

printfn "%i %i" p1.depth p1.horizontal

let p2 =
    input
    |> Array.fold
        (fun acc value ->
            match fst value with
            | "forward" ->
                { acc with
                    horizontal = acc.horizontal + snd value
                    depth = acc.depth + snd value * acc.aim }
            | "down" -> { acc with aim = acc.aim + snd value }
            | "up" -> { acc with aim = acc.aim - snd value }
            | _ -> acc)

        { horizontal = 0; depth = 0; aim = 0 }


printfn "%i %i" p2.depth p2.horizontal
